File: src/Functional/OCaml/patterns/command.ml
let () =
  let balance = ref 100 in
  let commands =
    [
      (fun () -> balance := !balance + 50);
      (fun () -> balance := !balance - 20);
    ]
  in
  List.iter (fun command -> command ()) commands;
  assert (!balance = 130)

File: src/Functional/Haskell/patterns/Command.hs
module Main where

data Command = Deposit Int | Withdraw Int deriving (Eq, Show)

execute :: Int -> Command -> Int
execute balance (Deposit amount) = balance + amount
execute balance (Withdraw amount) = balance - amount

main :: IO ()
main = do
  let queue = [Deposit 50, Withdraw 20]
      balance = foldl execute 100 queue
  if balance == 130 && length queue == 2
    then putStrLn "balance=130;commands=2"
    else error "Command contract failed"
